dotnet/src/Functions/Functions.Prompty/Extensions/PromptyKernelExtension.cs

```diff
@@ -0,0 +1,127 @@
+    /// <exception cref="ArgumentNullException"></exception>
+    /// <exception cref="NotSupportedException"></exception>
+    public static KernelFunction CreateFunctionFromPrompty(
+        this Kernel _,
+        string promptyPath,
+        IPromptTemplateFactory? promptTemplateFactory = null,
+        ILoggerFactory? loggerFactory = null)
+    {
+        var text = File.ReadAllText(promptyPath);
+
+        promptTemplateFactory ??= new LiquidPromptTemplateFactory(); // use liquid template factory by default
+
+        // create PromptTemplateConfig from text
+        // step 1
+        // retrieve the header, which is in yaml format and put between ---
+        //
+        // e.g
+        // file: chat.prompty
+        // ---
+        // name: Contoso Chat Prompt
+        // description: A retail assistant for Contoso Outdoors products retailer.
+        // authors:
+        //   - XXXX
+        // model:
+        //   api: chat
+        //   configuration:
+        //     type: azure_openai
+        //     azure_deployment: gpt - 35 - turbo
+        //     api_version: 2023 - 07 - 01 - preview
+        //   parameters:
+        //     tools_choice: auto
+        //     tools:
+        //     -type: function
+        //       function:
+        //         name: test
+        //         description: test function
+        //         parameters:
+        //           properties:
+        //             location:
+        //               description: The city and state or city and country, e.g.San Francisco, CA
+        //                 or Tokyo, Japan
+        // ---
+        // ... (rest of the prompty content)
+
+        var splits = text.Split(["---"], StringSplitOptions.RemoveEmptyEntries);
+        var yaml = splits[0];
+        var content = splits[1];
+
+        var deserializer = new DeserializerBuilder().Build();
+        var prompty = deserializer.Deserialize<PromptyYaml>(yaml);
+
+        // step 2
+        // create a prompt template config from the prompty object
+        var promptTemplateConfig = new PromptTemplateConfig
+        {
+            Name = prompty.Name, // TODO: sanitize name
+            Description = prompty.Description,
+            Template = content,
+        };
+
+        PromptExecutionSettings defaultExecutionSetting = prompty.Model?.ModelConfiguration?.ModelType switch
+        {
+            ModelType.azure_openai or ModelType.openai => new OpenAIPromptExecutionSettings()
+            {
+                ResponseFormat = prompty.Model?.Parameters?.ResponseFormat == "json_object" ? ChatCompletionsResponseFormat.JsonObject : null,
+                Temperature = prompty.Model?.Parameters?.Temperature ?? 1.0,
+                TopP = prompty.Model?.Parameters?.TopP ?? 1.0,
+                MaxTokens = prompty.Model?.Parameters?.MaxTokens,
+                Seed = prompty.Model?.Parameters?.Seed,
+                ModelId = prompty.Model?.ModelConfiguration?.AzureDeployment,
+            },
```

Comment:
@markwallace-microsoft, what would be the correct way to communicate to the service selector the desired service/model that a function wants to use? Do we already have a scheme for that?  Seems like we'd want to use whatever was specified in the prompty file as a hint to choose the best service of all that are registered.

Reply:
Right now we allow service id to be configured for a prompt like this

```json
{
     "schema": 1,
     "type": "completion",
     "description": "Creates a chat response to the user",
     "execution_settings": {
        "gpt-3.5-turbo": {
          "model_id": "gpt-3.5-turbo-0613",
          "max_tokens": 4000,
          "temperature": 0.1
        }
      },
     "input_variables": [
        {
          "name": "request",
          "description": "The user's request.",
          "required": true
        },
        {
          "name": "history",
          "description": "The history of the conversation.",
          "required": true
        }
     ]
}
```

At run time there must be a service with the specified id.

If a service id of "default" is provided then we can fall back to using that.

if multiple service ids are provided then the first match is used unless you provide a custom service selector.